File: novel-app/server/src/export.ts
// 导出模块，照搬 src-tauri/src/export.rs。
// 仅实现 md/txt（DOCX/PDF 由前端生成）。collect_export_data + render_markdown + render_plain_text。

import { listChaptersWithContent, readNovel, readOutline } from "./files/index.js";
import type { NovelData, Volume } from "./files/index.js";

// ExportData：Rust 用 #[serde(rename_all="camelCase")]，但 novel 是 NovelData（snake_case）
// 嵌套结构字段大小写混合；这里类型如实反映
export interface ExportChapter {
  num: number;
  title: string;
  words: number;
  body: string;
}

export interface ExportData {
  novel: NovelData;
  outline: Volume[];
  chapters: ExportChapter[];
}

// collect_export_data，照搬 export.rs:23-43
export function collectExportData(dir: string): ExportData {
  const novel = readNovel(dir);
  const outline = readOutline(dir);
  const rawChapters = listChaptersWithContent(dir);
  const chapters: ExportChapter[] = rawChapters.map(({ meta, body }) => ({
    num: meta.chapter,
    title: meta.title,
    words: meta.words,
    body,
  }));
  return { novel, outline, chapters };
}

// render_markdown，照搬 export.rs:45-91
export function renderMarkdown(data: ExportData): string {
  const parts: string[] = [];

  // 标题页
  parts.push(`# ${data.novel.title}\n`);
  parts.push(`> ${data.novel.genre} | ${data.novel.theme} | ${data.novel.target_chapters}章\n`);
  parts.push(`> 创建日期：${data.novel.created}\n`);

  if (data.novel.world) {
    parts.push("");
    parts.push("## 世界观\n");
    parts.push(data.novel.world);
  }

  if (data.novel.characters) {
    parts.push("");
    parts.push("## 角色\n");
    parts.push(data.novel.characters);
  }

  // 目录
  if (data.outline.length > 0) {
    parts.push("");
    parts.push("---\n");
    parts.push("## 目录\n");
    for (const volume of data.outline) {
      parts.push(`### ${volume.volume}\n`);
      for (const ch of volume.chapters) {
        parts.push(`- ${String(ch.num).padStart(3, "0")}. ${ch.title}`);
      }
      parts.push("");
    }
  }

  // 章节正文
  for (const chapter of data.chapters) {
    parts.push("");
    parts.push("---\n");
    parts.push(`# 第${chapter.num}章 ${chapter.title}\n`);
    parts.push(chapter.body);
  }

  return parts.join("\n");
}

// render_plain_text，照搬 export.rs:93-96
export function renderPlainText(data: ExportData): string {
  const md = renderMarkdown(data);
  return stripMarkdown(md);
}

// strip_markdown，照搬 export.rs:98-127
function stripMarkdown(text: string): string {
  let result = "";
  for (const line of text.split("\n")) {
    const trimmed = line.trim();
    let processed: string;
    if (trimmed === "---") {
      processed = "";
    } else if (trimmed.startsWith("# ")) {
      processed = `\n${stripFormatting(trimmed.slice(2))}\n`;
    } else if (trimmed.startsWith("## ")) {
      processed = `\n${stripFormatting(trimmed.slice(3))}\n`;
    } else if (trimmed.startsWith("### ")) {
      processed = `\n${stripFormatting(trimmed.slice(4))}\n`;
    } else if (trimmed.startsWith("> ")) {
      processed = `  ${stripFormatting(trimmed.slice(2))}`;
    } else if (trimmed.startsWith("- ")) {
      processed = `  · ${stripFormatting(trimmed.slice(2))}`;
    } else if (trimmed.startsWith("* ")) {
      processed = `  · ${stripFormatting(trimmed.slice(2))}`;
    } else {
      processed = stripFormatting(trimmed);
    }
    result += processed;
    result += "\n";
  }
  // 压缩 3+ 连续换行为 2 个
  while (result.includes("\n\n\n")) {
    result = result.replace(/\n\n\n/g, "\n\n");
  }
  return result.trim() + "\n";
}

// strip_formatting，照搬 export.rs:129-138
function stripFormatting(text: string): string {
  return stripSimpleCode(stripSimpleItalic(stripSimpleBold(text)));
}

// regex_simple_bold，照搬 export.rs:140-170
// 去除 **text** 的 ** 标记
function stripSimpleBold(text: string): string {
  const chars = [...text];
  let result = "";
  let i = 0;
  while (i < chars.length) {
    if (i + 1 < chars.length && chars[i] === "*" && chars[i + 1] === "*") {
      const start = i + 2;
      let end = start;
      while (end + 1 < chars.length) {
        if (chars[end] === "*" && chars[end + 1] === "*") break;
        end++;
      }
      if (end + 1 < chars.length) {
        result += chars.slice(start, end).join("");
        i = end + 2;
      } else {
        result += chars[i];
        i++;
      }
    } else {
      result += chars[i];
      i++;
    }
  }
  return result;
}

// regex_simple_italic，照搬 export.rs:172-201
// 去除 *text* 的 * 标记（不处理 **）
function stripSimpleItalic(text: string): string {
  const chars = [...text];
  let result = "";
  let i = 0;
  while (i < chars.length) {
    if (chars[i] === "*" && (i === 0 || chars[i - 1] !== "*")) {
      const start = i + 1;
      let end = start;
      while (end < chars.length) {
        if (chars[end] === "*" && (end + 1 >= chars.length || chars[end + 1] !== "*")) {
          break;
        }
        end++;
      }
      if (end < chars.length) {
        result += chars.slice(start, end).join("");
        i = end + 1;
      } else {
        result += chars[i];
        i++;
      }
    } else {
      result += chars[i];
      i++;
    }
  }
  return result;
}

// regex_simple_code，照搬 export.rs:203-229
// 去除 `text` 的 ` 标记
function stripSimpleCode(text: string): string {
  const chars = [...text];
  let result = "";
  let i = 0;
  while (i < chars.length) {
    if (chars[i] === "`") {
      const start = i + 1;
      let end = start;
      while (end < chars.length && chars[end] !== "`") {
        end++;
      }
      if (end < chars.length) {
        result += chars.slice(start, end).join("");
        i = end + 1;
      } else {
        result += chars[i];
        i++;
      }
    } else {
      result += chars[i];
      i++;
    }
  }
  return result;
}

// total_words，照搬 export.rs:231-233
export function totalWords(data: ExportData): number {
  return data.chapters.reduce((sum, c) => sum + c.words, 0);
}

File: app/src-tauri/src/export.rs
use std::path::Path;

use crate::error::Result;
use crate::files::{self, NovelData, Volume};

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ExportChapter {
    pub num: u32,
    pub title: String,
    pub words: u32,
    pub body: String,
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ExportData {
    pub novel: NovelData,
    pub outline: Vec<Volume>,
    pub chapters: Vec<ExportChapter>,
}

pub fn collect_export_data(dir: &Path) -> Result<ExportData> {
    let novel = files::read_novel(dir)?;
    let outline = files::read_outline(dir)?;
    let raw_chapters = files::list_chapters_with_content(dir)?;

    let chapters = raw_chapters
        .into_iter()
        .map(|(meta, body)| ExportChapter {
            num: meta.chapter,
            title: meta.title,
            words: meta.words,
            body,
        })
        .collect();

    Ok(ExportData {
        novel,
        outline,
        chapters,
    })
}

pub fn render_markdown(data: &ExportData) -> String {
    let mut parts = Vec::new();

    // 标题页
    parts.push(format!("# {}\n", data.novel.title));
    parts.push(format!(
        "> {} | {} | {}章\n",
        data.novel.genre, data.novel.theme, data.novel.target_chapters
    ));
    parts.push(format!("> 创建日期：{}\n", data.novel.created));

    if let Some(world) = &data.novel.world {
        parts.push(String::new());
        parts.push("## 世界观\n".to_string());
        parts.push(world.clone());
    }

    if let Some(characters) = &data.novel.characters {
        parts.push(String::new());
        parts.push("## 角色\n".to_string());
        parts.push(characters.clone());
    }

    // 目录
    if !data.outline.is_empty() {
        parts.push(String::new());
        parts.push("---\n".to_string());
        parts.push("## 目录\n".to_string());
        for volume in &data.outline {
            parts.push(format!("### {}\n", volume.volume));
            for ch in &volume.chapters {
                parts.push(format!("- {:03}. {}", ch.num, ch.title));
            }
            parts.push(String::new());
        }
    }

    // 章节正文
    for chapter in &data.chapters {
        parts.push(String::new());
        parts.push("---\n".to_string());
        parts.push(format!("# 第{}章 {}\n", chapter.num, chapter.title));
        parts.push(chapter.body.clone());
    }

    parts.join("\n")
}

pub fn render_plain_text(data: &ExportData) -> String {
    let md = render_markdown(data);
    strip_markdown(&md)
}

fn strip_markdown(text: &str) -> String {
    let mut result = String::new();
    for line in text.lines() {
        let trimmed = line.trim();
        let processed = if trimmed == "---" {
            String::new()
        } else if trimmed.starts_with("# ") {
            format!("\n{}\n", strip_formatting(&trimmed[2..]))
        } else if trimmed.starts_with("## ") {
            format!("\n{}\n", strip_formatting(&trimmed[3..]))
        } else if trimmed.starts_with("### ") {
            format!("\n{}\n", strip_formatting(&trimmed[4..]))
        } else if trimmed.starts_with("> ") {
            format!("  {}", strip_formatting(&trimmed[2..]))
        } else if trimmed.starts_with("- ") {
            format!("  · {}", strip_formatting(&trimmed[2..]))
        } else if trimmed.starts_with("* ") {
            format!("  · {}", strip_formatting(&trimmed[2..]))
        } else {
            strip_formatting(trimmed).to_string()
        };
        result.push_str(&processed);
        result.push('\n');
    }
    // 压缩连续空行
    while result.contains("\n\n\n") {
        result = result.replace("\n\n\n", "\n\n");
    }
    result.trim().to_string() + "\n"
}

fn strip_formatting(text: &str) -> String {
    let s = text.to_string();
    // 去除粗体 **text**
    let s = regex_simple_bold(&s);
    // 去除斜体 *text*
    let s = regex_simple_italic(&s);
    // 去除行内代码 `text`
    let s = regex_simple_code(&s);
    s
}

fn regex_simple_bold(text: &str) -> String {
    let mut result = String::new();
    let chars: Vec<char> = text.chars().collect();
    let mut i = 0;
    while i < chars.len() {
        if i + 1 < chars.len() && chars[i] == '*' && chars[i + 1] == '*' {
            // 找结束 **
            let start = i + 2;
            let mut end = start;
            while end + 1 < chars.len() {
                if chars[end] == '*' && chars[end + 1] == '*' {
                    break;
                }
                end += 1;
            }
            if end + 1 < chars.len() {
                for &c in &chars[start..end] {
                    result.push(c);
                }
                i = end + 2;
            } else {
                result.push(chars[i]);
                i += 1;
            }
        } else {
            result.push(chars[i]);
            i += 1;
        }
    }
    result
}

fn regex_simple_italic(text: &str) -> String {
    let mut result = String::new();
    let chars: Vec<char> = text.chars().collect();
    let mut i = 0;
    while i < chars.len() {
        if chars[i] == '*' && (i == 0 || chars[i - 1] != '*') {
            let start = i + 1;
            let mut end = start;
            while end < chars.len() {
                if chars[end] == '*' && (end + 1 >= chars.len() || chars[end + 1] != '*') {
                    break;
                }
                end += 1;
            }
            if end < chars.len() {
                for &c in &chars[start..end] {
                    result.push(c);
                }
                i = end + 1;
            } else {
                result.push(chars[i]);
                i += 1;
            }
        } else {
            result.push(chars[i]);
            i += 1;
        }
    }
    result
}

fn regex_simple_code(text: &str) -> String {
    let mut result = String::new();
    let chars: Vec<char> = text.chars().collect();
    let mut i = 0;
    while i < chars.len() {
        if chars[i] == '`' {
            let start = i + 1;
            let mut end = start;
            while end < chars.len() && chars[end] != '`' {
                end += 1;
            }
            if end < chars.len() {
                for &c in &chars[start..end] {
                    result.push(c);
                }
                i = end + 1;
            } else {
                result.push(chars[i]);
                i += 1;
            }
        } else {
            result.push(chars[i]);
            i += 1;
        }
    }
    result
}

pub fn total_words(data: &ExportData) -> u32 {
    data.chapters.iter().map(|c| c.words).sum()
}
